src/Service/RefundService.php

```diff
@@ -0,0 +1,333 @@
+use Shopware\Core\System\StateMachine\Exception\IllegalTransitionException;
+
+class RefundService
+{
+    const REFUNDABLE_STATES = [
+        OrderTransactionStates::STATE_PARTIALLY_REFUNDED,
+        OrderTransactionStates::STATE_PAID,
+        OrderTransactionStates::STATE_PARTIALLY_PAID,
+    ];
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var ConfigurationService
+     */
+    private $configurationService;
+
+    /**
+     * @var ClientService
+     */
+    private $clientService;
+
+    /**
+     * @var AdyenRefundRepository
+     */
+    private $adyenRefundRepository;
+
+    /**
+     * @var Currency
+     */
+    private $currency;
+
+    /** @var OrderTransactionRepository */
+    private $transactionRepository;
+
+    /** @var OrderTransactionStateHandler */
+    private $transactionStateHandler;
+
+    /**
+     * RefundService constructor.
+     *
+     * @param LoggerInterface $logger
+     * @param ConfigurationService $configurationService
+     * @param ClientService $clientService
+     * @param AdyenRefundRepository $adyenRefundRepository
+     * @param Currency $currency
+     * @param OrderTransactionRepository $transactionRepository
+     * @param OrderTransactionStateHandler $transactionStateHandler
+     */
+    public function __construct(
+        LoggerInterface $logger,
+        ConfigurationService $configurationService,
+        ClientService $clientService,
+        AdyenRefundRepository $adyenRefundRepository,
+        Currency $currency,
+        OrderTransactionRepository $transactionRepository,
+        OrderTransactionStateHandler $transactionStateHandler
+    ) {
+        $this->logger = $logger;
+        $this->configurationService = $configurationService;
+        $this->clientService = $clientService;
+        $this->adyenRefundRepository = $adyenRefundRepository;
+        $this->currency = $currency;
+        $this->transactionRepository = $transactionRepository;
+        $this->transactionStateHandler = $transactionStateHandler;
+    }
+
+    /**
+     * Process a refund on the Adyen platform
+     *
+     * @param OrderEntity $order
+     * @return array
+     * @throws AdyenException
+     */
+    public function refund(OrderEntity $order): array
+    {
+        $orderTransaction = $this->getAdyenOrderTransactionForRefund($order, self::REFUNDABLE_STATES);
+
+        // No param since sales channel is not available since we're in admin
+        $merchantAccount = $this->configurationService->getMerchantAccount();
+
+        if (!$merchantAccount) {
+            $message = 'No Merchant Account set. ' .
+                'Go to the Adyen plugin configuration panel and finish the required setup.';
+            $this->logger->error($message);
+            throw new AdyenException($message);
+        }
+
+        $pspReference = $orderTransaction->getCustomFields()[PaymentResponseHandler::ORIGINAL_PSP_REFERENCE];
```

Comment:
Add defensive coding here

Reply:
Validation is already done in the `getAdyenOrderTransactionForRefund` function call fore this line